### PR TITLE
Initial classes

### PR DESCRIPTION
**Pull request by mecha**
Initial classes for a basic event-driven state machine implementation.

The machine uses an event manager to trigger events on transition. A specialized `TransitionEvent` instance is given to the event manager. Handlers can use the event to abort transitions ~and also specify the new state of the machine.~ The transition key is used as the new state, unless the transition is aborted.

If an exception is thrown, the transition will still happen. Event handlers will need to catch exceptions and explicitly abort the transition.

### CODE REVIEW

File: src/AbstractEventStateMachine.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                $transition
+            );
+        }
+
+        if (!$event->isTransitionAborted()) {
+            $this->_setState($this->_getNewState($event));
+        }
+
+        if (isset($exception) && $exception instanceof Exception) {
+            throw $exception;
```
Comment (XedinUnknown) [inline src/AbstractEventStateMachine.php]: Nothing is apparently thrown if event handlers did not throw. However, the interface, AFAIK, [requires](https://github.com/Dhii/machine-state-machine-interface/blob/develop/src/StateMachineInterface.php#L31) one of two things happen: the transition happens, or [`CouldNotTransitionExceptionInterface`] is thrown. Therefore, the exception must be thrown when the transition is aborted, regardless of whether the event mechanism threw.

[`CouldNotTransitionExceptionInterface`]: https://github.com/Dhii/machine-state-machine-interface/blob/develop/src/Exception/CouldNotTransitionExceptionInterface.php

File: src/EventStateMachine.php [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+use Exception;
+use Psr\EventManager\EventManagerInterface;
+use RebelCode\State\Exception\CouldNotTransitionException;
+
+/**
+ * A readable, event-driven state machine.
+ *
+ * This implementation does not use an internal state graph. Instead, transition given to be applied will be
+ * used as the new state. As such, it is perfectly valid to transition to the same state, thus making this state
+ * machine implementation non-deterministic.
```
Comment (XedinUnknown) [inline src/EventStateMachine.php]: Looks deterministic to me.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]

File: src/EventStateMachine.php (hunk 2 above)
Comment (mecha) [reply src/EventStateMachine.php]: Forgot to remove after changing the mechanics. Thanks for pointing out.

**Review by XedinUnknown — APPROVED** [review]

File: src/AbstractEventStateMachine.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable|null $message  The error message, if any.
+     * @param int|null               $code     The error code, if any.
+     * @param Exception|null         $previous The previous exception for chaining, if any.
+     *
+     * @return CouldNotTransitionExceptionInterface The created exception.
+     */
+    abstract protected function _createStateMachineException(
```
Comment (XedinUnknown) [inline src/AbstractEventStateMachine.php:105]: Not used in this class. Why is it here?

**Review by XedinUnknown — CHANGES_REQUESTED** [review]

File: src/AbstractEventStateMachine.php (hunk 3 above)
Comment (mecha) [reply src/AbstractEventStateMachine.php:105]: Used in [`AbstractEventStateMachine#46`](https://github.com/RebelCode/event-state-machine/blob/864cbf673fe18f93ef97b775b8894f562c95c749/src/AbstractEventStateMachine.php#L46)

Comment (XedinUnknown) [reply src/AbstractEventStateMachine.php:105]: My bad.

**Review by XedinUnknown — APPROVED** [review]

**Pull request merged**